packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +460,347 @@ class ShellRoute extends RouteBase {
+    String? rootLocation,
+    this.name,
+    this.restorationScopeId,
+    this.preload = false,
+  })  : assert(rootLocation != null || (rootLocations?.isNotEmpty ?? false)),
+        rootLocations = rootLocations ?? <String>[rootLocation!],
+        navigatorKey = navigatorKey ??
+            GlobalKey<NavigatorState>(
+                debugLabel: name != null ? 'Branch-$name' : null);
+
+  /// The [GlobalKey] to be used by the [Navigator] built for this branch.
+  ///
+  /// A separate Navigator will be built for each StatefulShellBranch in a
+  /// [StatefulShellRoute] and this key will be used to identify the Navigator.
+  /// The routes associated with this branch will be placed o onto that
+  /// Navigator instead of the root Navigator.
   final GlobalKey<NavigatorState> navigatorKey;
+
+  /// The valid root locations for this branch.
+  final List<String> rootLocations;
+
+  /// An optional name for this branch.
+  final String? name;
+
+  /// Whether this route branch should be preloaded when the associated
+  /// [StatefulShellRoute] is visited for the first time.
+  ///
+  /// If this is true, this branch will be preloaded by navigating to
+  /// the root location (first entry in [rootLocations]).
+  final bool preload;
+
+  /// Restoration ID to save and restore the state of the navigator, including
+  /// its history.
+  final String? restorationScopeId;
+
+  /// Returns the default location for this branch (by default the first
+  /// entry in [rootLocations]).
+  String get defaultLocation => rootLocations.first;
+
+  /// Checks if this branch is intended to be used for the provided
+  /// GoRouterState.
+  bool isBranchFor(GoRouterState state) {
+    final String? match = rootLocations
+        .firstWhereOrNull((String e) => state.location.startsWith(e));
+    return match != null;
+  }
+
+  /// Gets the state for the current branch of the nearest stateful shell route
+  /// in the Widget tree.
+  static StatefulShellBranchState of(BuildContext context) =>
+      StatefulShellRoute.of(context).currentBranchState;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(other, this)) {
+      return true;
+    }
+    if (other is! StatefulShellBranch) {
+      return false;
+    }
+    return other.navigatorKey == navigatorKey &&
+        listEquals(other.rootLocations, rootLocations) &&
+        other.name == name &&
+        other.preload == preload &&
+        other.restorationScopeId == restorationScopeId;
+  }
+
+  @override
+  int get hashCode => Object.hash(
+      navigatorKey, rootLocations, name, preload, restorationScopeId);
+}
+
+/// StatefulShellRoute extension that provides support for resolving the
+/// current StatefulShellBranch.
+///
+/// Should not be used directly, consider using [StatefulShellRoute.of] or
+/// [StatefulShellBranch.of] to access [StatefulShellBranchState] for the
+/// current context.
+extension StatefulShellBranchResolver on StatefulShellRoute {
```

Comment:
This was to avoid polluting the public API with implementation details, but I guess the functionality of this extension could be moved to RouteBuilder.

Reply:
It would still show up in the public API, no?

Reply:
Well, since it isn't exposed (exported) in go_router.dart it wouldn't be visible when using the normal way to import go_router (`import 'package:go_router/go_router.dart';`). And as far as I understand it, dart doc, wouldn't generate documentation for it either.